### ISSUE TITLE
Refresh doesn't handle Statistics.db correctly?

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#1124]
In an application of "refresh", we saw tons of the following log messages:

```
[shard 6] sstable - Could not create SSTable component /var/lib/scylla data/redacted-ka-662-Statistics.db. Found exception: std::system_error (error system:17, File exists)
```

All of these error messages are on Statistics.db. Are we create this specific file or the whole sstable twice, or something silly like that?

**Comment by avikivity** [comment]
```
future<> sstable::mutate_sstable_level(uint32_t new_level) {
   if (!has_component(component_type::Statistics)) {
       return make_ready_future<>();
   }

   auto entry = _statistics.contents.find(metadata_type::Stats);
   if (entry == _statistics.contents.end()) {
       return make_ready_future<>();
   }

   auto& p = entry->second;
   if (!p) {
       throw std::runtime_error("Statistics is malformed");
   }
   stats_metadata& s = *static_cast<stats_metadata *>(p.get());
   if (s.sstable_level == new_level) {
       return make_ready_future<>();
   }

   s.sstable_level = new_level;
   // Technically we don't have to write the whole file again. But the assumption that
   // we will always write sequentially is a powerful one, and this does not merit an
   // exception.
   return seastar::async([this] {
       // This is not part of the standard memtable flush path, but there is no reason
       // to come up with a class just for that. It is used by the snapshot/restore mechanism
       // which comprises mostly hard link creation and this operation at the end + this operation,
       // and also (eventually) by some compaction strategy. In any of the cases, it won't be high
       // priority enough so we will use the default priority
       write_statistics(default_priority_class());
   });
}
future<> sstable::mutate_sstable_level(uint32_t new_level) {
   if (!has_component(component_type::Statistics)) {
       return make_ready_future<>();
   }

   auto entry = _statistics.contents.find(metadata_type::Stats);
   if (entry == _statistics.contents.end()) {
       return make_ready_future<>();
   }

   auto& p = entry->second;
   if (!p) {
       throw std::runtime_error("Statistics is malformed");
   }
   stats_metadata& s = *static_cast<stats_metadata *>(p.get());
   if (s.sstable_level == new_level) {
       return make_ready_future<>();
   }

   s.sstable_level = new_level;
   // Technically we don't have to write the whole file again. But the assumption that
   // we will always write sequentially is a powerful one, and this does not merit an
   // exception.
   return seastar::async([this] {
       // This is not part of the standard memtable flush path, but there is no reason
       // to come up with a class just for that. It is used by the snapshot/restore mechanism
       // which comprises mostly hard link creation and this operation at the end + this operation,
       // and also (eventually) by some compaction strategy. In any of the cases, it won't be high
       // priority enough so we will use the default priority
       write_statistics(default_priority_class());
   });
}
```

callled from ```load_new_sstables()```

**Comment by avikivity** [comment]
Exposed by

```
commit 1ff7d322725e15dd26ff48b735a16a498c732e17
Author: Raphael S. Carvalho <raphaelsc@scylladb.com>
Date:   Sat Mar 12 18:28:24 2016 -0300

    sstables: make write_simple() safer by using exclusive flag
    
    We should guarantee that write_simple() will not try to overwrite
    an existing file.
    
    Signed-off-by: Raphael S. Carvalho <raphaelsc@scylladb.com>
    Message-Id: <194bd055f1f2dc1bb9766a67225ec38c88e7b005.1457818073.git.raphaelsc@scylladb.com>

diff --git a/sstables/sstables.cc b/sstables/sstables.cc
index 176a46b..a55b459 100644
--- a/sstables/sstables.cc
+++ b/sstables/sstables.cc
@@ -895,7 +895,7 @@ template <sstable::component_type Type, typename T>
 void sstable::write_simple(T& component, const io_priority_class& pc) {
     auto file_path = filename(Type);
     sstlog.debug(("Writing " + _component_map[Type] + " file {} ").c_str(), file_path);
-    file f = new_sstable_component_file(file_path, open_flags::wo | open_flags::create | open_flags::truncate).get0();
+    file f = new_sstable_component_file(file_path, open_flags::wo | open_flags::create | open_flags::exclusive).get0();
 
     file_output_stream_options options;
     options.buffer_size = sstable_buffer_size;
```

**Comment by avikivity** [comment]
A fix would be to create the new statistics file under a new name, then rename it into place.

**Comment by avikivity** [comment]
Not only does mutate_sstable_level() recreate the statistics file, it does so on every shard!

**Comment by nyh** [comment]
We discussed this issue extensively on the mailing list, but for completeness, I want to explain the issue also here:

Leveled compaction (see https://github.com/scylladb/scylla/wiki/SSTable-compaction#leveled-compaction for description) assumes that the sstables held in one level have non-overlapping key ranges. So we can't import a new sstable directly to level 2, because its key range will likely overlap some of the existing level-2 sstables. So when importing an sstable, we must modify its level (which is strangely stored in the "statistics" component of the sstable) to be 0.

The existing code which tries to modify the sstable_level reads the content of the statistics file, modifies the sstable_level variable in the memory object, and writes the file back. The problem is that it writes it back using write_simple() which since a recent patch forbids overwriting, so this fails. It will only fail on sstables with sstable_level > 0 (otherwise the modification is not attempted), so this problem will only be seen when importing sstables coming from leveled compaction, which is apparently why we haven't noticed this bug in our tests.

Additionally, because we let all shards use all the sstables, the existing code does this modification in parallel on all shards, which is of course completely unnecessary and should be fixed.
